cpp/CMakeLists.txt

```diff
@@ -808,6 +808,7 @@ endif()
 add_subdirectory(src/arrow)
 
 if(ARROW_PARQUET)
+  add_definitions(-DARROW_PARQUET)
```

Comment:
In this PR, I'm experimenting with this define. I'm trying to fix https://github.com/apache/arrow/pull/6235/checks?check_run_id=457986026#step:4:635 Since after this PR `file_parquet.h` includes a parquet header (in particular `parquet_version.h` which is only generated when parquet is built), I've broken the independence of the parquet and dataset modules. This seemed like a potential fix but I think I'm going to remove the `#include` and copy paste/forward declare what I'm using in `file_parquet.h`

Re other components: I've added https://issues.apache.org/jira/browse/ARROW-7894 to address the problem more generally.